baseapp/baseapp.go

```diff
@@ -615,83 +620,62 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
+	// Attempt to execute all messages and only update state if all messages pass
+	// and we're in DeliverTx.
+	result, err = app.runMsgs(runMsgCtx, msgs, mode)
+	if err == nil && mode == runTxModeDeliver {
 		msCache.Write()
 	}
 
-	return result
+	return gInfo, result, err
 }
 
 // runMsgs iterates through all the messages and executes them.
-func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (result sdk.Result) {
+func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*sdk.Result, error) {
 	msgLogs := make(sdk.ABCIMessageLogs, 0, len(msgs))
-
 	data := make([]byte, 0, len(msgs))
-	var (
-		code      sdk.CodeType
-		codespace sdk.CodespaceType
-	)
-
 	events := sdk.EmptyEvents()
 
-	// NOTE: GasWanted is determined by ante handler and GasUsed by the GasMeter.
+	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
-		// match message route
+		// skip actual execution for (Re)CheckTx mode
+		if mode == runTxModeCheck || mode == runTxModeReCheck {
+			break
+		}
+
 		msgRoute := msg.Route()
 		handler := app.router.Route(msgRoute)
 		if handler == nil {
-			return sdk.ErrUnknownRequest("unrecognized message type: " + msgRoute).Result()
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized message route: %s", msgRoute)
 		}
 
-		var msgResult sdk.Result
-
-		// skip actual execution for CheckTx and ReCheckTx mode
-		if mode != runTxModeCheck && mode != runTxModeReCheck {
-			msgResult = handler(ctx, msg)
-		}
+		result, err := handler(ctx, msg)
 
 		// Each message result's Data must be length prefixed in order to separate
 		// each result.
-		data = append(data, msgResult.Data...)
-
-		msgEvents := msgResult.Events
+		data = append(data, result.Data...)
 
 		// append events from the message's execution and a message action event
+		msgEvents := result.Events
 		msgEvents = msgEvents.AppendEvent(
 			sdk.NewEvent(sdk.EventTypeMessage, sdk.NewAttribute(sdk.AttributeKeyAction, msg.Type())),
 		)
 
 		events = events.AppendEvents(msgEvents)
 
-		// stop execution and return on first failed message
-		if !msgResult.IsOK() {
-			msgLogs = append(msgLogs, sdk.NewABCIMessageLog(uint16(i), false, msgResult.Log, msgEvents))
-
-			code = msgResult.Code
-			codespace = msgResult.Codespace
+		// short-circuit if the message fails (i.e. disregard remaining messages)
+		if err != nil {
+			msgLogs = append(msgLogs, sdk.NewABCIMessageLog(uint16(i), false, result.Log, msgEvents))
 			break
 		}
 
-		msgLogs = append(msgLogs, sdk.NewABCIMessageLog(uint16(i), true, msgResult.Log, msgEvents))
-	}
-
-	result = sdk.Result{
-		Code:      code,
-		Codespace: codespace,
-		Data:      data,
-		Log:       strings.TrimSpace(msgLogs.String()),
-		GasUsed:   ctx.GasMeter().GasConsumed(),
-		Events:    events,
+		msgLogs = append(msgLogs, sdk.NewABCIMessageLog(uint16(i), true, result.Log, msgEvents))
 	}
 
-	return result
+	return &sdk.Result{
```

Comment:
This is odd... so if one msg returns an error, you update the logs and break processing (line 670), but still return err=nil to runTx. Make sure to return an error in that case. Otherwise, this will commit the prior messages (imagine msg 1 succeeded but msg 2 failed, this must return an error to cause the state changes from msg 1 to be reverted)

Reply:
Yes, this is a bug. See response in https://github.com/cosmos/cosmos-sdk/pull/5421#discussion_r360994196. Thanks for the review and catching this!